File: apps/server/internal/market/coinbase.go
package market

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
)

type CoinbaseProvider struct {
	Client  *http.Client
	BaseURL string
}

func (c *CoinbaseProvider) ID() string {
	return "coinbase"
}

func (c *CoinbaseProvider) FetchCandlesFromExchange(ctx context.Context, symbol string, start, end, granularity int64) ([]Candlestick, error) {
	// API Docs: /products/{product_id}/candles
	url := fmt.Sprintf("%s/products/%s/candles?granularity=%d&start=%d&end=%d",
		c.BaseURL, symbol, granularity, start, end)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", "Cochart-App")

	res, err := c.Client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(res.Body)
		return nil, fmt.Errorf("coinbase error %d: %s", res.StatusCode, string(body))
	}

	// Parse Response (Coinbase returns array of arrays)
	// [ [ time, low, high, open, close, volume ], ... ]
	var raw [][]float64
	if err := json.NewDecoder(res.Body).Decode(&raw); err != nil {
		return nil, fmt.Errorf("parse error: %v", err)
	}

	candles := make([]Candlestick, 0, len(raw))

	for _, c := range raw {
		if len(c) >= 6 {
			candles = append(candles, Candlestick{
				Timestamp: int64(c[0]),
				Low:       c[1],
				High:      c[2],
				Open:      c[3],
				Close:     c[4],
				Volume:    c[5],
			})
		}
	}

	// Coinbase returns newest first we usually want oldest first
	sort.Slice(candles, func(i, j int) bool {
		return candles[i].Timestamp < candles[j].Timestamp
	})

	return candles, nil
}

File: server/src/market/coinbase.ts
import type { Candlestick, ExchangeProvider, Product } from "./types";

const USER_AGENT = "Cochart-App";

interface CoinbaseProduct {
  id: string;
  quote_currency: string;
  base_currency: string;
  display_name: string;
  status: string;
}

export class CoinbaseProvider implements ExchangeProvider {
  constructor(private baseURL = "https://api.exchange.coinbase.com") { }

  id(): string {
    return "coinbase";
  }

  async getProducts(): Promise<Product[]> {
    const res = await fetch(`${this.baseURL}/products`, {
      headers: { "User-Agent": USER_AGENT },
    });

    if (!res.ok) {
      throw new Error(`coinbase status: ${res.status}`);
    }

    const raw = (await res.json()) as CoinbaseProduct[];

    const products: Product[] = [];
    for (const p of raw) {
      if (p.status !== "online") continue;
      products.push({
        id: p.id,
        name: p.base_currency + p.quote_currency,
        type: "crypto",
        exchange: "coinbase",
      });
    }
    return products;
  }

  async fetchCandles(
    symbol: string,
    start: number,
    end: number,
    granularity: number,
    signal?: AbortSignal,
  ): Promise<Candlestick[]> {
    // API Docs: /products/{product_id}/candles
    const url =
      `${this.baseURL}/products/${symbol}/candles` +
      `?granularity=${granularity}&start=${start}&end=${end}`;

    const res = await fetch(url, {
      headers: { "User-Agent": USER_AGENT },
      signal,
    });

    if (!res.ok) {
      const body = await res.text();
      throw new Error(`coinbase error ${res.status}: ${body}`);
    }

    // Coinbase returns an array of arrays:
    // [ [ time, low, high, open, close, volume ], ... ]
    const raw = (await res.json()) as number[][];

    const candles: Candlestick[] = [];
    for (const c of raw) {
      if (c.length >= 6) {
        candles.push({
          time: Math.trunc(c[0]),
          low: c[1],
          high: c[2],
          open: c[3],
          close: c[4],
          volume: c[5],
        });
      }
    }

    // Coinbase returns newest first; we want oldest first.
    candles.sort((a, b) => a.time - b.time);
    return candles;
  }
}
